app/src/main/kotlin/Twenty48/Game2048MDP.kt

```diff
@@ -4,41 +4,51 @@ import MDP
 
 class Game2048MDP(): MDP<Game2048State, Game2048Action>() {
 
+    private val gameObject = Game2048Controller() // We just use this for its methods
 
-    val gameObject = Game2048()
-    // val initialGrid = gameObject.grid.map{it.copyOf()}.toTypedArray().copyOf()
-    // val currentGameState = Game2048State(initialGrid, gameObject.gameScore.toDouble())
-    val initialGameState = Game2048State(gameObject.grid.map{it.copyOf()}.toTypedArray().copyOf() ) // Incorrect, the bug is the initial state is changing, it should remain fixed
-
+    val initialGrid = arrayOf(
+        arrayOf(0, 0, 0, 0),
+        arrayOf(0, 0, 0, 0),
+        arrayOf(0, 0, 0, 0),
+        arrayOf(0, 0, 0, 0)
+    )
+    val initialGamePosition = Game2048Position(initialGrid)
+    var initialGameState = Game2048State(initialGamePosition)
 
     override fun initialState(): Game2048State {
-        return this.initialGameState
+        return initialGameState // The MCTS solver is mutating the initial game state??
     }
 
     override fun reward(previousState: Game2048State?, action: Game2048Action?, state: Game2048State): Double {
-        // this.GameState.transitionAction(this.GameState.gameObject.manipulateGrid(this.grid, waitForValidInput()))
-        return this.gameObject.gameScore.toDouble()
+        return state.score!!.toDouble()
     }
 
     override fun transition(state: Game2048State, action: Game2048Action): Game2048State {
-        // this.gameObject.grid = this.gameObject.spawnNumber(this.gameObject.grid)
-        // this.gameObject.grid = this.gameObject.manipulateGrid(state.gameGrid, action.toString())
-        this.gameObject.run2048Programatic(this.gameObject.grid, action.toString())
-        // Fix issue with constant mutation, the state needs to be tied to the grid!!
-        // val newGrid = this.gameObject.run2048Programatic(this.gameObject.grid, action.toString())
-
-        // return Game2048State(newGrid, this.gameObject.gameScore.toDouble())
-        // gridworld score also
-        return Game2048State(gameObject.grid)
+
+        var tempGrid = gameObject.manipulateGrid(state.gameGrid, action.toString()) // this code mutates initialGameState
+        tempGrid = gameObject.spawnNumber(tempGrid)
+        val newGamePosition = Game2048Position(tempGrid)
```

Comment:
This code mutates the root stateNode in the solver for some reason. It no longer goes back to the original grid of all zeroes, but always continues from the last grid configuration. @JunTaoLuo